### DISCUSSION TITLE
Array2dEx internal data type

### DISCUSSION THREAD

**Discussion by cristianbuse** [Theadd/ds3x#25] [General]
Hi @Theadd ,

Is the goal of the ```Array2dEx``` class to always work with 2D Arrays of ```Variant``` or do you ever need to work with other data types?

I am only asking because at first glance it seems that you indeed have a need for ```Variant()``` only in which case there would be no need to wrap the ```Instance``` member in a ```Variant```. But if you do need flexibility and want to work with arrays of specific data types then things get more complex and even ```CloneParamArray``` needs to be re-thinked.

**Comment by Theadd** [comment]
Hi @cristianbuse ,

> Is the goal of the `Array2dEx` class to always work with 2D Arrays of `Variant` or do you ever need to work with other data types?

Well, when I decided to implement the `Array2dEx` class was to encapsulate a 2D Array within a class instance for various reasons:
- The simplicity to pass it back and forward between function calls as an object reference, avoiding to pass the whole array by value (yeah, a total failure until u pointed it out), by me or any other user.
- It had to be `Variant` or `Variant()` since it's main purpose back then was to copy the contents of  `ADODB.Recordset`s (with dates, strings, currencies, etc) to `Excel.Range` objects many, many times faster than setting cell values through iterating the range. Also allowing me to transform the data before copying it. 
- The straightforward `Array2dEx.Create()` factory.
- And **immutability**/**pure functions**.

I've been improving `Array2dEx` leveraging your `VBA-MemoryTools` techniques. [First](https://github.com/Theadd/ds3x/blob/5329ce7e7064e7b6bf7241f5097de9d900aea007/src/ds3x.accdb.src/modules/MemoryLib.cls#L934-L959), converting the `Array2dEx.CloneArray(): CloneParamArray` call so it copies the requested range directly to its corresponding position within the 2d array being reconstructed in the `ArraySliceGroup` instance instead of creating intermediate arrays in each `ArraySlice`. 
And [later](https://github.com/Theadd/ds3x/blob/5329ce7e7064e7b6bf7241f5097de9d900aea007/src/ds3x.accdb.src/modules/Array2dEx.cls#L62-L70), since the `Instance As Variant` problem was also causing the whole array being **assigned** by value, or just being deallocated when the declaring scope ends (if assigned using `GetArrayByRef`), I declared a `Private pInstance() As Variant` as companion, pointed `Instance` to `pInstance()` and reassigned the safearray to them, removing it from the provided array to prevent deallocation. 

In other words (simplified):
```vb
Public Instance As Variant
Private pInstance() As Variant

Public Sub SetInstanceByRef(ByRef TargetArray() As Variant)
    Redim pInstance(0 To 0, 0 To 0)
    Instance = GetArrayByRef(pInstance)
    MemLongPtr(VarPtrArr(Instance)) = MemLongPtr(VarPtrArr(TargetArray))
    MemLongPtr(VarPtrArr(TargetArray)) = NULL_PTR
End Sub
```

It might not be the best approach tho,

**Comment by cristianbuse** [comment]
Hi @Theadd ,

Apologies, I should have been more explicit.

As you already know, an array or type ```Variant``` (e.g. ```Dim arr() As Variant```) can hold any data type per element because each and every element is a ```Variant``` wrapper.

What I actually wanted to discuss was arrays of differend data types e.g. ```Dim arr() As Double``` where each element of the array can only be of a specific data type. If you really need to accomodate arrays of other data types, then the internal data must be wrapped in a ```Variant```. Here's why:

```VBA
Option Explicit

Sub TestArrayTypes()
    Dim arrV() As Variant: ReDim arrV(0 To 1)
    Dim arrD() As Double:  ReDim arrD(0 To 1)
    Dim arrL() As Long:    ReDim arrL(0 To 1)
    Dim v As Variant:      ReDim v(0 To 1) 'Defaults to 'As Variant'
    Dim w As Variant:      ReDim w(0 To 1) As Long
    '
    arrV = v    'Works because both arrays are Variant()
    arrV = w    'Runtime error
    arrV = arrD 'Compiler error
    arrV = arrL 'Compiler error
    arrD = arrL 'Compiler error
    arrD = w    'Runtime error
    arrL = arrD 'Compiler error
    arrL = arrV 'Compiler error
    arrL = w    'Works because both arrays are Long()
    v = arrV    'Works because v is Variant and can be assigned any value except custom data types
    v = arrD    'Works because v is Variant
    v = arrL    'Works because v is Variant
    v = w       'Works because v is Variant
End Sub
```

So, as you can see, if you need to deal with anything else than ```Variant()``` then you must store the data wrapped inside a ```Variant```.

However, if you only ever need ```Variant()``` then you should not wrap into a ```Variant``` because of the following performance impact.
Running this:

```VBA
Option Explicit

Sub TestPerformanceImpact()
    Dim arr() As Variant
    Dim v As Variant
    Dim r As Variant
    Dim t As Double
    Dim i As Long
    Dim j As Long
    Const iterations As Long = 1000000
    Dim lowB As Long: lowB = 0
    Dim uppB As Long: uppB = 99
    '
    ReDim arr(lowB To uppB)
    For i = lowB To uppB
        Select Case i Mod 3
            Case 0: arr(i) = Rnd() * i
            Case 1: arr(i) = Space$(i)
            Case 2: arr(i) = i * 17
        End Select
    Next i
    v = arr
    '
    t = Timer
    For j = 1 To iterations
        For i = lowB To uppB
            r = arr(i)
        Next i
    Next j
    Debug.Print "Reading items from 'Variant()' takes (seconds): " & Round(Timer - t, 3)
    '
    t = Timer
    For j = 1 To iterations
        For i = lowB To uppB
            r = v(i)
        Next i
    Next j
    Debug.Print "Reading items from 'Variant' takes (seconds):   " & Round(Timer - t, 3)
End Sub
```
prints this:
![image](https://github.com/Theadd/ds3x/assets/23198997/0d1d4400-d712-4b2a-8d77-82bf7fcee226)

If the array is wrapped in a ```Variant``` then reading the array elements can be slower depending on the array size and the data stored inside. That's because an extra redirection is done by VB behind the scene.

Something else to consider is that you can pass a ```v() As Variant``` to a method expecting a ```ByRef v As Variant``` but you cannot pass an array wrapped in a ```Variant``` to a method expecting a ```ByRef v() As Variant```.

**Comment by Theadd** [comment]
Hi @cristianbuse and thanks for all the feedback, I'm not even able to keep it up to date. =)

> However, if you only ever need `Variant()` then you should not wrap into a `Variant`

I totally agree, I was planning to update the code accordingly after introducing the `Private pInstance() As Variant`, which I just did https://github.com/Theadd/ds3x/commit/3d358d896383b5e7fa7c9a9bd95c216f83989035 although, for now, only within the `Array2dEx` class, I'll get back to it as soon as possible.

I've also been using the `Variant()` data type right away in the drop-in replacement for the **mscorlib** `ArrayList` that I've been working on in the past few days, taking advantage of pointers and direct in-memory copy whenever possible. Finally dropping the *423MB VMem* dependency on `mscorlib` with pretty nice *and totally unexpected* performance boost. <em><sup>(7X times faster on .Add calls, 10X on .ToArray, 13X accessing, etc.)</sup></em>
[ArrayList.cs](https://github.com/Theadd/ds3x/blob/main/src/ds3x.accdb.src/modules/ArrayList.cls) 

Also, I've been looking at your [VBA-WeakReference](https://github.com/cristianbuse/VBA-WeakReference) project and it's really interesting, I already have a few places in mind that would totally benefit from it, in the "live editor" and at work. In fact, it's a neat and reliable method to manage and relate child form hierarchies, simplifying props and state propagation.
